mm2src/mm2_event_stream/src/manager.rs

```diff
@@ -0,0 +1,451 @@
+    /// laziness of `*_or_else()` functions).
+    ///
+    /// `data_fn` will only be evaluated if the streamer is found and accepts an input.
+    pub fn send_fn<T: Send + 'static>(
+        &self,
+        streamer_id: &str,
+        data_fn: impl FnOnce() -> T,
+    ) -> Result<(), StreamingManagerError> {
+        let streamers = self.streamers.read();
+        let streamer_info = streamers
+            .get(streamer_id)
+            .ok_or(StreamingManagerError::StreamerNotFound)?;
+        let data_in = streamer_info.data_in.as_ref().ok_or(StreamingManagerError::NoDataIn)?;
+        data_in
+            .unbounded_send(Box::new(data_fn()))
+            .map_err(|e| StreamingManagerError::SendError(e.to_string()))
+    }
+
+    /// Stops streaming from the streamer with `streamer_id` to the client with `client_id`.
+    pub fn stop(&self, client_id: u64, streamer_id: &str) -> Result<(), StreamingManagerError> {
+        let mut clients = self.clients.lock().unwrap();
+        if let Some(listening_to) = clients.get_mut(&client_id) {
+            listening_to.remove(streamer_id);
+
+            let mut streamers = self.streamers.write();
+            streamers
+                .get_mut(streamer_id)
+                .ok_or(StreamingManagerError::StreamerNotFound)?
+                .clients
+                .remove(&client_id);
+
+            // If there are no more listening clients, terminate the streamer.
+            if streamers.get(streamer_id).map(|info| info.clients.len()) == Some(0) {
+                streamers.remove(streamer_id);
+            }
+            Ok(())
+        } else {
+            Err(StreamingManagerError::UnknownClient)
+        }
+    }
+
+    /// Broadcasts some event to clients listening to it.
+    ///
+    /// In contrast to `StreamingManager::send`, which sends some data to a streamer,
+    /// this method broadcasts an event to the listening *clients* directly, independently
+    /// of any streamer (i.e. bypassing any streamer).
+    pub fn broadcast(&self, event: Event) {
+        if let Some(client_ids) = self.streamers.read().get(event.origin()).map(|info| &info.clients) {
+            self.controller.read().broadcast(event, Some(client_ids))
+        }
+    }
+
+    /// Forcefully broadcasts an event to all known clients even if they are not listening for such an event.
+    pub fn broadcast_all(&self, event: Event) { self.controller.read().broadcast(event, None) }
+
+    pub fn new_client(&self, client_id: u64) -> Result<Receiver<Arc<Event>>, StreamingManagerError> {
+        let mut clients = self.clients.lock().unwrap();
+        if clients.contains_key(&client_id) {
+            return Err(StreamingManagerError::ClientExists);
+        }
+        clients.insert(client_id, HashSet::new());
+        // Note that events queued in the channel are `Arc<` shared.
+        // So a 1024 long buffer isn't actually heavy on memory.
+        Ok(self.controller.write().create_channel(client_id, 1024))
+    }
```

Comment:
The comment is only justifying why `1024` entries aren't some large number memory-wise (since these are `arc`s)
but we still need to bound size the channel since this the directly being consumed in the SSE push code and controlled by external clients.
if we were to have this unbounded and a client is slow for whatever reason, this can grow unbounded.

aside from the channel size itself being able to grow unbounded (albeit, a hard to trigger vulnerability since we store light `arc<`s), the underlying events won't get freed since there is at least one remaining reference is still alive (the `arc<`s in the channel).

for example if the client is very slow we might have 5k pending events, `arc<`ed ofc, but the actual 5k events are still in memory and can't be removed yet since that one client isn't ingesting them.

in here i assume that if a client is that slow to receive the events (1024 pending), we better off drop correctness for more safety.